libvast/src/evaluate.cpp

```diff
@@ -0,0 +1,518 @@
+      const auto row = detail::narrow_cast<int64_t>(id - offset);
+      if (!array.IsNull(row))
+        continue;
+      result.append(false, id - result.size());
+      result.append(true, 1u);
+    }
+    result.append(false, offset + array.length() - result.size());
+    return result;
+  }
+};
+
+// Special-case not-equal operations with nil.
+template <concrete_type LhsType>
+struct column_evaluator<relational_operator::not_equal, LhsType, caf::none_t> {
+  static ids
+  evaluate([[maybe_unused]] LhsType type, id offset, const arrow::Array& array,
+           [[maybe_unused]] caf::none_t rhs, const ids& selection) noexcept {
+    // TODO: This entire loop semantically is just selection &
+    // array.null_bitmap, but we cannot do bitwise operations with Arrow's
+    // bitmaps yet.
+    ids result{};
+    for (auto id : select(selection)) {
+      VAST_ASSERT(id >= offset);
+      const auto row = detail::narrow_cast<int64_t>(id - offset);
+      if (array.IsNull(row))
+        continue;
+      result.append(false, id - result.size());
+      result.append(true, 1u);
+    }
+    result.append(false, offset + array.length() - result.size());
+    return result;
+  }
+};
+
+// All operations comparing with nil except for equal and not-equal always yield
+// no results.
+template <relational_operator Op, concrete_type LhsType>
+struct column_evaluator<Op, LhsType, caf::none_t> {
+  static ids
+  evaluate([[maybe_unused]] LhsType type, id offset, const arrow::Array& array,
+           [[maybe_unused]] caf::none_t rhs,
+           [[maybe_unused]] const ids& selection) noexcept {
+    return ids{offset + array.length(), false};
+  }
+};
+
+// For operations comparing enumeration arrays with a string view we want to
+// first convert the string view into its underlying integral representation,
+// and then dispatch to that column evaluator.
+template <relational_operator Op>
+struct column_evaluator<Op, enumeration_type, view<std::string>> {
+  static ids
+  evaluate(enumeration_type type, id offset, const arrow::Array& array,
+           view<std::string> rhs, const ids& selection) noexcept {
+    if (auto key = type.resolve(rhs)) {
+      auto rhs_internal = detail::narrow_cast<view<enumeration>>(*key);
+      return column_evaluator<Op, enumeration_type, view<enumeration>>::evaluate(
+        type, offset, array, rhs_internal, selection);
+    }
+    return ids{offset + array.length(), false};
+  }
+};
+
+// A utility function for evaluating meta extractors in predicates. This is
+// always a yes or no question per batch, so the function does not have to deal
+// with bitmaps at all.
+bool evaluate_meta_extractor(const table_slice& slice,
+                             const meta_extractor& lhs, relational_operator op,
+                             const data& rhs) {
+  switch (lhs.kind) {
+    case meta_extractor::kind::type: {
+      switch (op) {
+#define VAST_EVAL_DISPATCH(op)                                                 \
```

Comment:
Couldn't this be a templated function instead of a macro?

Reply:
Not without reflection or a lot of __FUNCTION_NAME__ magic à la [magic_enum](https://github.com/Neargye/magic_enum/).